frame/clearing-house/src/lib.rs

```diff
@@ -126,28 +144,84 @@ pub mod pallet {
 
 	// Pallets use events to inform users when important changes are made.
 	#[pallet::event]
-	// #[pallet::generate_deposit(pub(super) fn deposit_event)]
-	pub enum Event<T: Config> {}
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Margin successfully added to account
+		MarginAdded {
+			/// Account id that received the deposit
+			account: T::AccountId,
+			/// Asset type deposited
+			asset: AssetIdOf<T>,
+			/// Amount of asset deposited
+			amount: T::Balance,
+		},
+	}
 
 	// ----------------------------------------------------------------------------------------------------
 	//                                           Runtime  Errors
 	// ----------------------------------------------------------------------------------------------------
 
 	// Errors inform users that something went wrong.
 	#[pallet::error]
-	pub enum Error<T> {}
+	pub enum Error<T> {
+		/// User attempted to deposit unsupported asset type as collateral in its margin account
+		UnsupportedCollateralType,
+	}
 
 	// ----------------------------------------------------------------------------------------------------
 	//                              Extrinsics
 	// ----------------------------------------------------------------------------------------------------
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		/// Adds margin to a user's account. A user has to have enough margin to open new positions
+		/// and can be liquidated if its margin ratio falls bellow maintenance. Deposited collateral
+		/// backs all the positions of an account accross multiple markets (cross-margining).
+		///
+		/// If an account does not exist in `AccountsMargin`, it is created and initialized with 0
+		/// margin. Checks that the collateral type is supported.
+		#[pallet::weight(0)]
+		pub fn add_margin(
+			origin: OriginFor<T>,
+			asset: AssetIdOf<T>,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let acc = ensure_signed(origin)?;
+			<Self as MarginAccounts>::add_margin(&acc, asset, amount)?;
+			Ok(())
+		}
+	}
 
 	// ----------------------------------------------------------------------------------------------------
 	//                              Trait Implementations
 	// ----------------------------------------------------------------------------------------------------
 
+	impl<T: Config> MarginAccounts for Pallet<T> {
+		type AccountId = T::AccountId;
+		type AssetId = AssetIdOf<T>;
+		type Balance = T::Balance;
+
+		fn add_margin(
+			account: &Self::AccountId,
+			asset: Self::AssetId,
+			amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			ensure!(
+				CollateralTypes::<T>::contains_key(asset),
+				Error::<T>::UnsupportedCollateralType
+			);
+
+			// Assuming stablecoin collateral and all markets quoted in dollars
+			T::Assets::transfer(asset, account, &T::PalletId::get().into_account(), amount, true)?;
```

Comment:
Open to suggestions about whether or not to set the `keep_alive` option to `true`